src/Nethermind/Nethermind.Evm/CodeAnalysis/StatsAnalyzer.cs

```diff
@@ -0,0 +1,201 @@
+        const ulong fivegramBitMask = (255UL << 8 * 4) | fourgramBitMask;
+        const ulong sixgramBitMask = (255UL << 8 * 5) | fivegramBitMask;
+        const ulong sevengramBitMask = (255UL << 8 * 6) | sixgramBitMask;
+        public ulong[] ngramBitMaks = [255UL, twogramBitMask, threegramBitMask, fourgramBitMask, fivegramBitMask, sixgramBitMask, sevengramBitMask];
+        public static ulong[] byteIndexes = {
+            255UL,
+            255UL << 8,
+            255UL << 16,
+            255UL << 24,
+            255UL << 32,
+            255UL << 40,
+            255UL << 48,
+            255UL << 56
+        };
+        public static ulong[] byteIndexShifts = {
+             0,
+             8,
+             16,
+             24,
+             32,
+             40,
+             48,
+             56
+        };
+
+        private StatsAnalyzer(uint topN, string statFileName, string dir = "/media/usb/stats/")
+        {
+            _dir = dir;
+            _n = topN;
+            _fullPath = _dir + statFileName + ".stats.txt";
+            _sketch = new CMSketch(2, 600000);
+            _topN = new SortedSet<(ulong ngram, uint count)>(new TopNComparer());
+            _topNMap = new Dictionary<ulong, uint>();
+            SetStatFileName(statFileName);
+        }
+
+        public static StatsAnalyzer GetInstance(uint topN, string statFileName, string dir = "/media/usb/stats/")
+        {
+            lock (_lock)
+            {
+                if (_instance == null)
+                {
+                    _instance = new StatsAnalyzer(topN, statFileName, dir);
+                }
+                else
+                {
+                    _instance.SetStatFileName(statFileName);
+                }
+                return _instance;
+            }
+        }
+
+        public void SetInstanceStatFileName(string statFileName)
+        {
+            lock (_lock)
+            {
+                if (_instance != null)
+                {
+                    _instance.SetStatFileName(statFileName);
+                }
+            }
+        }
+
+        public void SetStatFileName(string statFileName)
+        {
+
+            _fullPath = _dir + statFileName + ".stats.txt";
+            // Ensure the log file is empty or create a new one
+            if (File.Exists(_fullPath))
+            {
+                File.WriteAllText(_fullPath, string.Empty);  // Clear the file content
+            }
+        }
+
+        public static void AddInstruction(Instruction instruction)
+        {
+            lock (_lock)
+            {
+                if (_instance != null)
+                {
+                    _instance.AddByte((byte)instruction);
+                }
+            }
+        }
+
+        public void AddByte(byte instruction)
+        {
+            _ngram = (_ngram << 8) | instruction;
+
+            for (int i = 1; i < 7; i++)
+            {
+                if (byteIndexes[i] < _ngram) AddSequence(_ngram & ngramBitMaks[i]);
```

Comment:
Ah, this is smart! Why track separately if we can track the biggest ngram and mask it properly. 🧠